autofit/graphical/mean_field.py

```diff
@@ -260,6 +323,62 @@ def from_dist(
     ) -> "MeanField":
         return dist if isinstance(dist, cls) else MeanField(dist)
 
+    def update_factor_mean_field(
+        self,
+        cavity_dist: "MeanField",
+        last_dist: Optional["MeanField"] = None,
+        delta: float = 1.0,
+        status: Status = Status(),
+    ) -> Tuple["MeanField", Status]:
+
+        success, messages, _, flag = status
+        updated = False
+        try:
+            with LogWarnings(logger=_log_projection_warnings, action='always') as caught_warnings:
+                factor_dist = self / cavity_dist
+                if delta < 1:
+                    log_norm = factor_dist.log_norm
+                    factor_dist = factor_dist ** delta * last_dist ** (1 - delta)
+                    factor_dist.log_norm = (
+                        delta * log_norm + (1 - delta) * last_dist.log_norm
+                    )
+
+            for m in caught_warnings.messages:
+                messages += (f"project_mean_field warning: {m}",)
+
+            if not factor_dist.is_valid:
+                success = False
+                messages += (f"model projection for {self} is invalid",)
+                factor_dist = factor_dist.update_invalid(last_dist)
+                # May want to check another way
+                # e.g. factor_dist.check_valid().sum() / factor_dist.check_valid().size
+                valid = factor_dist.check_valid()
+                if valid.any():
+                    updated = True
+                    n_valid = valid.sum()
+                    n_total = valid.size
+                    logger.debug(
+                        "meanfield with variables: %r ,"
+                        "partially updated %d parameters "
+                        "out of %d total, %.0%%",
```

Comment:
You can do formatting with fstrings

Reply:
When logging it's better to use c string formatting as then the interpolation only happens when the logging occurs.

https://docs.python.org/3/howto/logging.html#logging-variable-data

Reply:
That link is broken for some reason. 503 so I think the server is down. *shakes fist at Russia*

That seems to make sense for debug level logging in a part of the code that gets called a lot. I suspect for the most part string formatting for logs has little impact on execution time so I'd go for clarity over speed.

Reply:
Is it really that much less readable?

Reply:
To my mind it is because you have to cross-match to see where variables go